### PR TITLE
Remove identifiers from definitions, allow nested definitions instead

### PR DESCRIPTION
**Pull request by mnapoli**
This was discussion via hangout today. The idea is that `DefinitionProviderInterface` should return definitions indexed by definition IDs. This makes much more sense, and is much more optimized for real scenarios (even though it's an implementation detail, it's a major detail if we have to call `getIdentifier` on 100+ objects every single time).

That also allows definitions to be nested (since they have no names anymore), instead of having to expose them through the container via a definition name. This is the equivalent of "private services" in Symfony for example, or nested definitions in PHP-DI.

That had the effect of making alias definitions and references identical => the reference became a definition and aliases are not needed anymore. I don't know if it's obvious, feedback welcome.

### CODE REVIEW

**Comment by Anahkiasen** [comment]
I wrote a **lot** of definitions today and can confirm that it feels very redundant to define identifiers twice everywhere. Moreover, currently one would  instinctively think the array key is the identifier – I know because that's how I thought it worked in the first place. So, :+1: for me

**Comment by mnapoli** [comment]
Also a note on why I didn't remove `DefinitionInterface`: while I don't like unnecessary code, the interface is used in many type-hints/docblocks. This is even more used with this change because references have been replaces by definitions. If there is no `DefinitionInterface`, then there is no way to describe "definition objects" in the documentation and the code… I could be convinced otherwise but having the interface made more sense when doing the PR.

**Comment by Anahkiasen** [comment]
No yeah I agree on keeping the interface nonetheless, can be useful to assert that a given variable is a definition without necessarily the need to assert that it can do anything particular. Just, to be sure we didn't receive a scalar or something else.

File: src/AliasDefinitionInterface.php [hunk 1]
```diff
@@ -1,10 +0,0 @@
-<?php
-
-namespace Interop\Container\Definition;
-
-/**
- * Alias a container entry to another.
- */
-interface AliasDefinitionInterface extends DefinitionInterface, ReferenceInterface
-{
-}
```
Comment (Anahkiasen) [inline src/AliasDefinitionInterface.php:10]: If I understand correctly to define an alias you would now do per example:

``` php
[
  'identifier' => $definition,
  'alias' => new Reference('identifier'),
]
```

Correct?

Comment (mnapoli) [inline src/AliasDefinitionInterface.php:10]: Yep!

Comment (Anahkiasen) [inline src/AliasDefinitionInterface.php:10]: Nice.

**Comment by moufmouf** [comment]
I'd like to try implementing this first in Yaco and YamlDefinitionLoader before merging (just to see how it feels). Do you mind if we wait one day or two before merging?

**Comment by mnapoli** [comment]
No problem for me!

**Comment by moufmouf** [comment]
Ok, I've been doing some tests in Yaco and YamlDefinitionLoader, and this change seems to definitely make a lot of sense. Code is cleaner, less redundant, and I faced absolutely no problem while implementing it.

Merging now!

**Pull request merged**